lib/publiccloud/ssh_interactive.pm

```diff
@@ -1,67 +1,93 @@
+
     # Prepare the environment for the SSH tunnel
     my $upload_port = get_required_var('QEMUPORT') + 1;
     my $upload_host = testapi::host_ip();
 
-    $instance->ssh_script_run(
-        # Create /dev/sshserial fifo on remote and tail|tee it to /dev/$serialdev on local
-        #   timeout => switches to script_run instead of script_output to be used so the test will not wait for the command to end
-        #   tunnel the worker port (for downloading from data/ and uploading assets / logs
-        cmd => "'rm -rf /dev/sshserial; mkfifo -m a=rwx /dev/sshserial; tail -fn +1 /dev/sshserial' 2>&1 | tee /dev/$serialdev; clear",
-        timeout => 0,
-        no_quote => 1,
-        ssh_opts => "-yt -R $upload_port:$upload_host:$upload_port",
-        username => 'root',
-    );
-    sleep 3;
-    save_screenshot;
+    # Pipe the output of the device fifo to the local serial terminal
+# Note: We run this in a loop so that the ssh tunnel gets automatically re-established after device reboots and such. The sleep helps to avoid unnecessary CPU hogging in case of connection issues
+    enter_cmd("while true; do ssh sut -yt -R '$upload_port:$upload_host:$upload_port' 'rm -f /dev/sshserial && mkfifo -m a=rwx /dev/sshserial && tail -fn +1 /dev/sshserial' 2>&1 >/dev/$serialdev; sleep 5; done");
+    # give the ssh connection some time to settle
+    sleep 10;
 
+    # from here onwards, the serial output should be directed to /dev/sshserial instead to /dev/ttyS0
     set_var('SERIALDEV_', $serialdev);
+    set_sshserial_dev();
     set_var('_SSH_TUNNELS_INITIALIZED', 1);
-
     set_var('AUTOINST_URL_HOSTNAME', 'localhost');
-    set_sshserial_dev();
+
+    # When re-activing the consoles we also need to setup the underlying ssh connections again,
+    # because this only happens the first time the console is activated
+    if ($args{reconnect}) {
+        establish_tunnel_console("root-console");
+        establish_tunnel_console("user-console");
+        establish_tunnel_console("root-virtio-terminal");
+    }
+
+    select_console($prev_console) if ($prev_console !~ /tunnel-console/);
 }
 
 sub ssh_interactive_leave {
-    # Check if the SSH tunnel is still up and leave the SSH interactive session
-    script_run("test -p /dev/sshserial && exit", timeout => 0);
-
-    # Restore the environment to not use the SSH tunnel for upload/download from the worker
-    #set_var('SUT_HOSTNAME',          testapi::host_ip());
-    set_var('AUTOINST_URL_HOSTNAME', testapi::host_ip());
+    my $prev_console = current_console();
+    # Switch to the local console, terminate the ssh tunnel and set the serial device back
+    select_console('tunnel-console') unless ($prev_console =~ /tunnel-console/);
     unset_sshserial_dev();
+    set_var('AUTOINST_URL_HOSTNAME', testapi::host_ip());
+
+    # Terminate the ssh tunnel loop. Send ctrl-c twice to ensure it arrives
+    send_key 'ctrl-c';
```

Comment:
Change 2/2 - Terminate the foreground process by sending a CTRL+C. I'm sending it twice so it will be fore sure terminated. This has no negative impact.